packages/xy-chart/src/utils/computeCirclePack.js

```diff
@@ -1,7 +1,271 @@
+            let isect = 0;
+            let s1 = 1;
+            let s2 = 1;
+            for (j = b.pack_next; j !== b; j = j.pack_next, s1 += 1) {
+              if (xpackIntersects(j, c)) {
+                isect = 1;
+                break;
+              }
+            }
+            if (isect === 1) {
+              for (k = a.pack_prev; k !== j.pack_prev; k = k.pack_prev, s2 += 1) {
+                if (xpackIntersects(k, c)) {
+                  break;
+                }
+              }
+            }
+            // update front chain
+            if (isect) {
+              if (s1 < s2 || (s1 === s2 && b.r < a.r)) {
+                xpackSplice(a, b = j);
+              } else {
+                xpackSplice(a = k, b);
+              }
+              i -= 1;
+              preiter = true;
+            } else {
+              xpackInsert(a, c);
+              b = c;
+              bound(c, bd);
+              preiter = false;
+            }
+          }
+        }
+      }
+    }
+    return nodes.length;
+  }
+
+  if (data.length === 0) {
+    return data;
+  }
+
+  const nodes = [];
+  data.forEach((node) => {
+    nodes.push(Object.assign({}, node, { x: xScale(node.x), r: node.size }));
+  });
+
+  for (let i = 0; i < nodes.length; i += 1) {
+    xpackLink(nodes[i]);
+  }
+
+  let i = 0;
+  let n = 0;
+  while (i < nodes.length) {
+    packBounds.push({
+      xMin: Infinity,
+      yMin: Infinity,
+      xMax: -Infinity,
+      yMax: -Infinity,
+    });
+
+      // pack this nodes group
+    i = packNodes(nodes, i, n);
+
+    // get the packing outline
+    packOutline.push(nodes[i - 1]);
+    for (let p = nodes[i - 1].pack_next; p !== nodes[i - 1]; p = p.pack_next) {
+      packOutline.push(p);
+    }
+    n += 1;
+  }
+
+  rect.xMin = Infinity;
+  rect.xMax = -Infinity;
+  rect.yMin = Infinity;
+  rect.yMax = -Infinity;
+  packBounds.forEach((bd) => {
+    rect.xMin = Math.min(rect.xMin, bd.xMin);
+    rect.xMax = Math.max(rect.xMax, bd.xMax);
+    rect.yMin = Math.min(rect.yMin, bd.yMin);
+    rect.yMax = Math.max(rect.yMax, bd.yMax);
+  });
+
+        // compute the global outline
+  packOutline.sort((a, b) => (a.px - b.px));
+  packOutline.forEach((p) => {
+    if (p.py < 0) {
+      outline.up.push(p);
+    } else if (p.py > 0) {
+      outline.down.push(p);
+    } else {
+      outline.up.push(p);
+      outline.down.push(p);
+    }
+  });
+  return nodes;
+}
+
+export default function computeCirclePack(data, xScale, yScale) {
   const sorted = data.sort((a, b) => a.x - b.x);
-  return sorted.map(d => ({
-    ...d,
-    y: ((1 - d.importance) * (Math.random() > 0.5 ? 1 : -1)),
-  }));
+  const calculatedNodes = packCircles(data, xScale);
+  const result = [];
+  for (let i = 0; i < calculatedNodes.length; i += 1) {
+    result.push(Object.assign(
+      {},
+      sorted[i],
+      {
+        x: xScale.invert(calculatedNodes[i].px),
+        y: calculatedNodes[i].py,
+      }));
+  }
+  const modifiedYScale = yScale.copy();
```

Comment:
it seems like this functionality could potentially live in the `CirclePackSeries` component since it doesn't really have to do with circle packing. this would make the logic / functions easier to test as well.